cni/network/network_linux_test.go

```diff
@@ -160,3 +168,387 @@ func TestAddSnatForDns(t *testing.T) {
+							{
+								Family: platform.AfINET,
+								// matches cns ip configuration (20.241.0.0/16 == 20.241.0.35/16)
+								Prefix: *getIPNetWithString("20.241.0.0/16"),
+								// matches cns ip configuration gateway ip address
+								Gateway: net.ParseIP("20.241.0.35"),
+							},
+						},
+					},
+					epIDRegex: `.*`,
+				},
+				// should match secondary
+				{
+					epInfo: &network.EndpointInfo{
+						MacAddress:  parsedMACAddress,
+						ContainerID: "test-container",
+						Data: map[string]interface{}{
+							"vethname": "nettest-containereth0",
+						},
+						Routes: []network.RouteInfo{
+							{
+								Dst: *getCIDRNotationForAddress("169.254.2.1/32"),
+								Gw:  net.ParseIP("10.244.2.1"),
+							},
+							{
+								Dst: *getCIDRNotationForAddress("0.0.0.0/0"),
+								Gw:  net.ParseIP("169.254.2.1"),
+							},
+						},
+						PODName:           "test-pod",
+						PODNameSpace:      "test-pod-ns",
+						NICType:           cns.NodeNetworkInterfaceFrontendNIC,
+						SkipDefaultRoutes: false,
+						MasterIfName:      "secondary",
+						NetworkID:         "net",
+						NetNsPath:         "bc526fae-4ba0-4e80-bc90-ad721e5850bf",
+						NetNs:             "bc526fae-4ba0-4e80-bc90-ad721e5850bf",
+						HostSubnetPrefix:  "<nil>",
+						EndpointDNS: network.DNSInfo{
+							Servers: []string{
+								"ns1", "ns2",
+							},
+							Suffix: "myDomain",
+						},
+						Options: map[string]interface{}{
+							"testflag": "copy",
+						},
+						// matches with cns ip configuration
+						IPAddresses: []net.IPNet{
+							{
+								IP:   net.ParseIP("10.241.0.35"),
+								Mask: getIPNetWithString("10.241.0.35/32").Mask,
+							},
+						},
+						NATInfo: nil,
+						// ip config pod ip + mask(s) from cns > interface info > subnet info
+						Subnets: []network.SubnetInfo{
+							{
+								Family: platform.AfINET,
+								Prefix: *getIPNetWithString("10.241.0.35/32"),
+								// matches cns ip configuration gateway ip address
+								Gateway: net.ParseIP("10.241.0.35"),
+							},
+						},
+					},
+					epIDRegex: `.*`,
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			err := tt.plugin.Add(tt.args)
+			require.NoError(t, err)
+			allEndpoints, _ := tt.plugin.nm.GetAllEndpoints("")
+			require.Len(t, allEndpoints, len(tt.want))
+
+			// compare contents
+			for _, wantedEndpointEntry := range tt.want {
+				epID := "none"
+				for _, endpointInfo := range allEndpoints {
+					if !tt.match(wantedEndpointEntry.epInfo, endpointInfo) {
+						continue
+					}
+					// save the endpoint id before removing it
+					epID = endpointInfo.EndpointID
+					require.Regexp(t, regexp.MustCompile(wantedEndpointEntry.epIDRegex), epID)
+
+					// omit endpoint id and ifname fields as they are nondeterministic
+					endpointInfo.EndpointID = ""
+					endpointInfo.IfName = ""
+
+					require.Equal(t, wantedEndpointEntry.epInfo, endpointInfo)
```

Comment:
Why do we compare twice? Why do we omit these fields and then compare?

Reply:
The tt.match is to identify which item in the map we are looking at, so that we compare the full endpoint info struct to the correct item in the map. The map ordering is nondeterministic.

We omit the endpoint id and interface name because they are based on which interface info is processed first in the map, and what interface info is processed first is nondeterministic since it is a map.